### ISSUE TITLE
[Feature] Read icons from system cache

### ISSUE THREAD

**Issue by Hofknecht** [Hofknecht/SystemTrayMenu#149]
There should be a better user experience if we decrease the time for first load of a folder.
(Most time is I/O, also there it takes more time e.g. the first time we access the drive).
During first time we open a folder, we read the icon's into a iconcache (accessing icon is time consuming), which is not yet saved and restored each restart.
The reason for this is that a icon can change, but in practise it does not very often, so it may be an improvement when we store it into our app cache and only reading it in background, then only if necessary update the gui

the disadvantage of this solution is, that there will remain icon's in this cache which are no longer needed if e.g. file deleted?

**Comment by topeterk** [comment]
Windows already has an icon cache.
Is STM bypassing this cache or is it just not working?

**Comment by Hofknecht** [comment]
not sure, what i see is that this method to get a icon takes ~3ms (from our DictIconCache nearly instant <1ms), even if it is called second time. i assume it is using windows cache, but no idea

IntPtr hImageList = DllImports.NativeMethods.Shell32SHGetFileInfo(
                filePath,
                DllImports.NativeMethods.FileAttributeNormal,
                ref shfi,
                (uint)Marshal.SizeOf(shfi),
                flags);

i remember we tried a lot to get really every icon and also that it works in every resolution

not sure if we can replace this or if it is the right way to store that own DictIconCache 

needs further investigation

**Comment by Hofknecht** [comment]
https://stackoverflow.com/questions/968369/fastest-way-to-get-the-extension-name-and-icon-by-an-file-extension?rq=1
"Using SHGetFileInfo is more likely to work on every operating system version and will be compatible with however Microsoft do things in future"

=> so probably we can not reduce these ~3ms per icon. So DictIconCache is still good solution to access same icons faster.
But to store these tmp icons is probably bad solution

maybe we can reduce the times by new settings feature to preload specifc submenus.
At the moment we preload only main menu. 

but for now i will close topic here because my root idea here is not usefull

**Comment by topeterk** [comment]
There are multiple methods getting an icon as suggested here:
https://stackoverflow.com/questions/35497503/how-to-get-system-image-list-icon-index-of-an-ishellitem

I would suggest trying [IShellItemImageFactory::GetImage](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ishellitemimagefactory-getimage) to get the icon from the system icon cache. At least try getting the in-memory icon first and if this fails load the icon afterwards async, as suggested by Microsoft:

> _Icon extraction can be time consuming. This method generally should not be called from a UI thread to avoid causing that thread to become unresponsive. You can call IShellItemImageFactory::GetImage on a UI thread if you set the SIIGBF_INCACHEONLY flag. However, if the image is not found in the cache, the calling application should be prepared to launch a background thread to extract the image. An extraction should never be done on a UI thread._

It might not solve loading times completely, as the documentation also says (whenever disk must be accessed to acquire the icon) but for all in-memory icons this **could** be faster (if SHGetFileInfo isn't already using the system icon cache).

Maybe give it a shot?

**Comment by Hofknecht** [comment]
thank you for that hint.

i tried, but it seems the images which are returned are not that what we are looking for:
![image](https://user-images.githubusercontent.com/52528841/115283808-27aa7100-a14c-11eb-8b2b-23d15975ce28.png)
(maybe i'm doing something wrong)

Interresting is also that here the best case (when image cached) is  ≤ 4 ms  (current Shell32SHGetFileInfo is  ≤ 3ms )
![image](https://user-images.githubusercontent.com/52528841/115283613-e6b25c80-a14b-11eb-8b41-7bcebf50354f.png)

**Comment by topeterk** [comment]
I guess you got the thumbnail (the preview of an image file, for example) but you can force getting the icon instead:
> SIIGBF_ICONONLY (0x00000004)
Return only the icon, never the thumbnail.

**Comment by Hofknecht** [comment]
thank you! works!

there is only a problem left with transparent icons 

current (left side: windows exploer, right side: stm):
![image](https://user-images.githubusercontent.com/52528841/115459996-015a0380-a228-11eb-94c7-04892d2b79f4.png)
![image](https://user-images.githubusercontent.com/52528841/115460076-1a62b480-a228-11eb-8b03-d301d79f9dce.png)


new: (left side: windows exploer, right side: stm):
![image](https://user-images.githubusercontent.com/52528841/115460034-1040b600-a228-11eb-9e9f-49c93b10c1e9.png)
![image](https://user-images.githubusercontent.com/52528841/115460085-1cc50e80-a228-11eb-8edb-4c8de2e6729c.png)


For most as you can see in the screenshots the transparent is also ok, but some not

new code at the moment:
```
public static Icon GetIcon(string filename)
        {
            Icon icon = null;

            IShellItem ppsi = null;

            // GUID of IShellItem.
            Guid uuid = new Guid("43826d1e-e718-42ee-bc55-a1e261c37bfe");

            try
            {
                SHCreateItemFromParsingName(filename, IntPtr.Zero, uuid, out ppsi);
                IntPtr hbitmap = IntPtr.Zero;
                ((IShellItemImageFactory)ppsi).GetImage(new SIZE(16, 16), SIIGBF.SIIGBF_ICONONLY, out hbitmap);
                Bitmap image = Image.FromHbitmap(hbitmap);
                DeleteObject(hbitmap);

                image.MakeTransparent();
                IntPtr icH = image.GetHicon();
                icon = Icon.FromHandle(icH);
                image.Dispose();
            }
#pragma warning disable CA1031 // Do not catch general exception types
            catch (Exception ex)
#pragma warning restore CA1031 // Do not catch general exception types
            {
                if (ex is FileNotFoundException)
                {
                    Log.Warn($"filename:'{filename}'", ex);
                }
                else
                {
                    Log.Error($"filename:'{filename}'", ex);

                    // throw; // comment in to remove CA1031
                }
            }

            return icon;
        }
```

+
NativeDllImport/DeleteObject.cs
NativeDllImport/GetImage.cs
NativeDllImport/IShellItem.cs
NativeDllImport/IShellItemImageFactory.cs
NativeDllImport/SHCreateItemFromParsingName.cs
NativeDllImport/SIGDN.cs
NativeDllImport/SIIGBF.cs
NativeDllImport/SIZE.cs
(from http://pinvoke.net/default.aspx/Interfaces/IShellItem.html?diff=y)

**Comment by topeterk** [comment]
There are image conversions involved. Maybe there is anything wrong with the alpha channel?

```
Bitmap image = Image.FromHbitmap(hbitmap);
image.MakeTransparent();
icon = Icon.FromHandle(icH);
```

It feels like ```Bitmap::MakeTransparent``` is maybe using the wrong color or the alpha channel gets already lost at ```Image::FromHbitmap```. If this is the reason, I found [this one](https://mycsharp.de/forum/threads/68883/geloest-alphakanal-geht-bei-bitmap-fromhbitmap-verloren?page=1#forumpost-3527354) that "more or less" seems to solve this issue (more like a workaround but hey, a good start if it works and is fast enough).

**Comment by Hofknecht** [comment]
works, performance is similar ~ 3ms 
also no gdi leaks from testing
comitted to master

**Comment by Hofknecht** [comment]
@topeterk 
i will close, reopen if you have any concerns 


other points related to this topic:

1
only for explanation related to the object DictIconCache, we still caching icons for standard application these are all extentions except    extensionsWithDiffIcons  ".EXE", ".LNK", ".ICO", ".URL"  and folders
so that we read per standard application only once, that makes performance much faster expecially in my big list test where i have thousends of txt files in it loading nearly instant
hope that point is ok in your eyes

2
probably we have also to remove TAFactory.IconPack.ExtractAllIcons because as i can see it has a gdi leak. for current version it is very low because it is only for a special read of a *.sln (so you would need to open 5.000 times a folder including a *sln until stm crashes through it, but is still a issue, so i created extra bug for it:
[BUG] Replace TAFactory.IconPack.ExtractAllIcons #155
 
if you have any idea to replace 
```
// e.g. VS 2019 icon, need another icom in imagelist
                List<Icon> extractedIcons = IconHelper.ExtractAllIcons(
                    executable.ToString());
                icon = extractedIcons.Last();
                handled = true;
```

**Comment by Hofknecht** [comment]
oh i just see there is still a issue with linkoverlay icons 
before we made it like described here
https://stackoverflow.com/questions/2701263/get-the-icon-for-a-given-extension

but i think we can workaround this by 
```
public static Icon AddIconOverlay(Icon originalIcon, Icon overlay)
        {
            Icon icon = null;
            if (originalIcon != null)
            {
                using Bitmap target = new Bitmap(
                    originalIcon.Width,
                    originalIcon.Height,
                    PixelFormat.Format32bppArgb);
                Graphics graphics = Graphics.FromImage(target);
                graphics.DrawIcon(originalIcon, 0, 0);
                graphics.DrawIcon(overlay, 0, 0);
                target.MakeTransparent(target.GetPixel(1, 1));
                icon = Icon.FromHandle(target.GetHicon());
            }

            return icon;
        }
```
therefore we would need the linkoverlay icon in solution i will try this out

**Comment by Hofknecht** [comment]
i played around with it. Seems AddIconOverlay has always a GDI leak problem (and also maybe a performance issue), even if changing to:
```
                static Icon AddIconOverlay(Icon originalIcon, Icon overlay)
                {
                    Icon icon = null;
                    if (originalIcon != null)
                    {
                        using Bitmap target = new Bitmap(
                            originalIcon.Width,
                            originalIcon.Height,
                            PixelFormat.Format32bppArgb);
                        {
                            using (Graphics graphics = Graphics.FromImage(target))
                            {
                                graphics.DrawIcon(originalIcon, 0, 0);
                                graphics.DrawIcon(overlay, 0, 0);
                            }

                            target.MakeTransparent(target.GetPixel(1, 1));
                            icon = (Icon)Icon.FromHandle(target.GetHicon()).Clone();
                            NativeMethods.User32DestroyIcon(target.GetHicon());
                            NativeMethods.User32DestroyIcon(originalIcon.Handle);
                        }

                        //originalIcon.Dispose();
                        originalIcon = null;
                    }

                    return icon;
                }
```

 For 'Icon overlay' i putted a Linkoverlay.ico into resources.

I searched in google a bit, found
https://docs.microsoft.com/en-us/windows/win32/shell/how-to-implement-icon-overlay-handlers
https://docs.microsoft.com/en-us/windows/win32/shell/how-to-create-icon-handlers
but there is no code example, but maybe advantage would be that we can show e.g. git status as link overlay as well when we would implement that, not sure 


From testing i assume SHGetFileInfo is already using system cache too because it has very simlar performance and also after restart it is caching as fast as the new one, so i consider to revert changes from #149

**Comment by Hofknecht** [comment]
https://github.com/Hofknecht/SystemTrayMenu/releases
in version 1.0.17.18 new method,  link overlay not working, but we can test/compare beahvior, so we still have the code and a release version if further needed (e.g. compare performance)
in version 1.0.17.19 reverted, link overlay working again

i will close this issue for now, dont hestiate to reopen or open a new and link it, if any further concerns
